File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // the logic is pretty clear, check which half is sorted and do a bs there


        int l=0;
        int r=nums.size()-1;

        while(l<=r)
        {
            int mid=l+(r-l)/2;

            if(nums[mid]==target) return mid;

            if(nums[mid]>=nums[l])
            {
                if(nums[l]<=target && target<nums[mid])
                {
                    r=mid-1;
                }
                else
                {
                    l=mid+1;
                }
            }

            else
            {
                if(nums[mid]<target && nums[r]>=target)
                {
                    l=mid+1;
                }
                else
                {
                    r=mid-1;
                }
            }
        }

        return -1;

        
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # now basically in this we know one thing that one part is always sorted and target les in that part
        # once we find that part we can do a binary search on it


        # okay so how can we do it

        left=0
        right=len(nums)-1

        while left<=right:
            mid=(left+right)//2

            if nums[mid]==target:
                return mid

            elif nums[left]<=nums[mid]:
                if nums[left]<=target<nums[mid]:
                    right=mid-1
                else:
                    left=mid+1
            else:
                if nums[mid]<target<=nums[right]:
                    left=mid+1
                else:
                    right=mid-1

        return -1
        